### PR TITLE
feat: add VITA-Audio training

### PR DESCRIPTION
**Pull request by weedge**
feat:
- add VITA-Audio training with modal
- add remove train output dir with modal

> [!TIP]
> 运行脚本可以将数据处理单独出来， 通过多模态 tokenizer (encode) 将数据转换成文本token， 文本LLM的tokenizer 需要添加这些特殊的token,  通过preprocess data 预处理成token 存放， 以便 直接 流式批量训练 LLM with mtp

### CODE REVIEW

**Comment by weedge** [comment]
# just debug | 2xA100-80GB for 7B with DeepSpeed ZeRO 1/2
```shell
IMAGE_GPU=A100-80GB:2 MAX_STEPS=1 modal run src/train/vita_audio/finetune_glm4voice_stage1.py --other-args "--dataset_joint true --dataloader_num_workers 8"
```
image

train loss and train metrics:
```
[default0]:{'train_runtime': 216.6525, 'train_samples_per_second': 0.148, 'train_steps_per_second': 0.005, 'train_loss': 13.93506908416748, 'epoch': 1.0}
[default0]:***** train metrics *****
[default0]:  epoch                    =        1.0
[default0]:  total_flos               =   652587GF
[default0]:  train_loss               =    13.9351
[default0]:  train_runtime            = 0:03:36.65
[default0]:  train_samples            =          2
[default0]:  train_samples_per_second =      0.148
[default0]:  train_steps_per_second   =      0.005
```

**Comment by weedge** [comment]
# stage2: Freeze LLM , training Single MCTP Module | just debug | 2xA100-80GB for 7B with DeepSpeed ZeRO 1/2
```shell
IMAGE_GPU=A100-80GB:2 MAX_STEPS=1 modal run src/train/vita_audio/finetune_glm4voice_mtp1_stage1.py
```
image

1xA100-80G也可以用于debug, 但是为了测试多卡的情况。

train loss and train metrics:
```
[default0]:{'train_runtime': 297.6561, 'train_samples_per_second': 0.108, 'train_steps_per_second': 0.003, 'train_loss': 13.0, 'epoch': 1.0}
[default0]:***** train metrics *****
[default0]:  epoch                    =        1.0
[default0]:  total_flos               =   676276GF
[default0]:  train_loss               =       13.0
[default0]:  train_runtime            = 0:04:57.65
[default0]:  train_samples            =          2
[default0]:  train_samples_per_second =      0.108
[default0]:  train_steps_per_second   =      0.003
```
```
- GF (GigaFLOPS) = 10^9 浮点运算/秒
- TFLOPS (TeraFLOPS) = 10^12 浮点运算/秒
所以 1 TFLOPS = 1000 GF

652587GF - 这是之前训练日志中的 total_flos，表示训练过程中执行的总浮点运算数量，而不是每秒的运算速度。

训练运行时间是 297.6561 秒。

训练过程中的平均计算速度是：

676276GF ÷ 297.6561秒 ≈ 2272.0 GF/s = 2.272 TFLOPS
```

**Comment by weedge** [comment]
# stage3: Freeze LLM , training Multiple MCTP Modules(10) | just debug | 2xA100-80GB for 7B with DeepSpeed ZeRO 1/2
```shell
IMAGE_GPU=A100-80GB:2 MAX_STEPS=1 modal run src/train/vita_audio/finetune_glm4voice_mtp1_stage1.py
```
image

train loss and train metrics:
```
[default0]:{'train_runtime': 333.3402, 'train_samples_per_second': 0.096, 'train_steps_per_second': 0.003, 'train_loss': 12.75, 'epoch': 1.0}
[default0]:***** train metrics *****
[default0]:  epoch                    =        1.0
[default0]:  total_flos               =   889484GF
[default0]:  train_loss               =      12.75
[default0]:  train_runtime            = 0:05:33.34
[default0]:  train_samples            =          2
[default0]:  train_samples_per_second =      0.096
[default0]:  train_steps_per_second   =      0.003
```
```
- GF (GigaFLOPS) = 10^9 浮点运算/秒
- TFLOPS (TeraFLOPS) = 10^12 浮点运算/秒
所以 1 TFLOPS = 1000 GF

889484GF - 这是之前训练日志中的 total_flos，表示训练过程中执行的总浮点运算数量，而不是每秒的运算速度。

训练运行时间是 333.3402 秒。

训练过程中的平均计算速度是：

676276GF ÷ 333.3402 秒 ≈ 2668.4 GF/s = 2.6684 TFLOPS
```

**Comment by weedge** [comment]
# stage4: Supervised Fine-tuning  LLM  and Multiple MCTP Modules(10) | just debug | 2xA100-80GB for 7B with DeepSpeed ZeRO 2 no optimizer with offload to cpu
image

```shell
IMAGE_GPU=A100-80GB:2 MAX_STEPS=1 modal run src/train/vita_audio/finetune_glm4voice_mtp10_stage2.py
```
image

train loss and train metrics:
```
[default0]:{'train_runtime': 433.7106, 'train_samples_per_second': 0.074, 'train_steps_per_second': 0.002, 'train_loss': 26.755382537841797, 'epoch': 1.0}
[default0]:***** train metrics *****
[default0]:  epoch                    =        1.0
[default0]:  total_flos               =   889484GF
[default0]:  train_loss               =    26.7554
[default0]:  train_runtime            = 0:07:13.71
[default0]:  train_samples            =          2
[default0]:  train_samples_per_second =      0.074
[default0]:  train_steps_per_second   =      0.002
```
```
- GF (GigaFLOPS) = 10^9 浮点运算/秒
- TFLOPS (TeraFLOPS) = 10^12 浮点运算/秒
所以 1 TFLOPS = 1000 GF

889484GF - 这是之前训练日志中的 total_flos，表示训练过程中执行的总浮点运算数量，而不是每秒的运算速度。

训练运行时间是 433.7106 秒。

训练过程中的平均计算速度是：

676276GF ÷ 433.7106 秒 ≈ 2050.9 GF/s = 2.0509 TFLOPS
```

stage4 和 stage3 一样的GPU 浮点运算数量; LLM 部分计算offload to cpu

**Comment by weedge** [comment]
```shell
tensorboard --logdir /train_output/finetune_glm4voice_stage1/runs
tensorboard --logdir /train_output/finetune_glm4voice_mtp1_stage1/runs
tensorboard --logdir /train_output/finetune_glm4voice_mtp10_stage1/runs
tensorboard --logdir /train_output/finetune_glm4voice_mtp10_stage2/runs
```

**Comment by weedge** [comment]
# evaluate datasets
from huggingface datasets, need convert to evaluate format datasets(jsonl)

## speech -> text QA
- https://huggingface.co/datasets/fixie-ai/llama-questions
- https://huggingface.co/datasets/fixie-ai/trivia_qa-audio
- https://huggingface.co/datasets/fixie-ai/spoken-web-questions

## speech -> text ASR
- https://huggingface.co/datasets/fixie-ai/librispeech_asr
- https://huggingface.co/datasets/wenet-e2e/wenetspeech
- https://huggingface.co/datasets/shenyunhang/AISHELL-1

## text -> speech TTS
- https://huggingface.co/datasets/mythicinfinity/libritts
- https://github.com/BytedanceSpeech/seed-tts-eval
